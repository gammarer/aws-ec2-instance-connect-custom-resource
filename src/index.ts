import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId, PhysicalResourceIdReference } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

export interface EC2InstanceConnectCustomResourceProps {
  readonly subnetId: string;
  readonly securityGroupIds: string[];
}

export class EC2InstanceConnectCustomResource extends AwsCustomResource {

  constructor(scope: Construct, id: string, props: EC2InstanceConnectCustomResourceProps) {
    super(scope, id, {
      installLatestAwsSdk: true,
      onUpdate: {
        service: 'EC2',
        action: 'createInstanceConnectEndpoint',
        parameters: {
          SubnetId: props.subnetId,
          DryRun: false,
          PreserveClientIp: true,
          SecurityGroupIds: props.securityGroupIds,
        },
        physicalResourceId: PhysicalResourceId.fromResponse('InstanceConnectEndpoint.InstanceConnectEndpointId'),
      },
      onDelete: {
        service: 'EC2',
        action: 'deleteInstanceConnectEndpoint',
        parameters: {
          DryRun: false,
          InstanceConnectEndpointId: new PhysicalResourceIdReference(),
        },
      },
      policy: AwsCustomResourcePolicy.fromStatements([
        new PolicyStatement({
          actions: [
            'iam:CreateServiceLinkedRole',
            'ec2:CreateInstanceConnectEndpoint',
            'ec2:CreateNetworkInterface',
            'ec2:CreateTags',
            'ec2:DeleteInstanceConnectEndpoint',
          ],
          resources: ['*'],
        }),
      ]),
    });
  }
}