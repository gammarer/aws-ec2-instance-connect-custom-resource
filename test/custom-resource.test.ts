import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { AwsCustomResource } from 'aws-cdk-lib/custom-resources';
import { EC2InstanceConnectCustomResource } from '../src';

describe('CustomResource Testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const iccr = new EC2InstanceConnectCustomResource(stack, 'EC2InstanceConnectCustomResource', {
    subnetId: 'example-subnet-id',
    securityGroupIds: [
      'example-security-group-id',
    ],
  });

  const template = Template.fromStack(stack);

  it('Is CustomResource', () => {
    expect(iccr).toBeInstanceOf(AwsCustomResource);
  });

  it('Should have Lambda Function Role', () => {
    template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
      AssumeRolePolicyDocument: Match.objectEquals({
        Version: '2012-10-17',
        Statement: Match.arrayWith([
          Match.objectEquals({
            Effect: 'Allow',
            Principal: {
              Service: 'lambda.amazonaws.com',
            },
            Action: 'sts:AssumeRole',
          }),
        ]),
      }),
      ManagedPolicyArns: Match.arrayWith([
        {
          'Fn::Join': Match.arrayEquals([
            '',
            Match.arrayEquals([
              'arn:',
              {
                Ref: 'AWS::Partition',
              },
              ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
            ]),
          ]),
        },
      ]),
    }));
  });

  it('Should have Lambda Function', () => {
    template.hasResourceProperties('AWS::Lambda::Function', Match.objectEquals({
      Code: {
        S3Bucket: Match.anyValue(),
        S3Key: Match.stringLikeRegexp('.*.zip'),
      },
      Handler: 'index.handler',
      Role: {
        'Fn::GetAtt': [
          Match.stringLikeRegexp('.*ServiceRole.*'),
          'Arn',
        ],
      },
      Runtime: 'nodejs14.x',
      Timeout: 120,
    }));
  });

  it('Should have Custom Resource Policy', () => {
    template.hasResourceProperties('AWS::IAM::Policy', Match.objectEquals({
      PolicyName: Match.anyValue(),
      PolicyDocument: {
        Statement: [
          {
            Action: [
              'iam:CreateServiceLinkedRole',
              'ec2:CreateInstanceConnectEndpoint',
              'ec2:CreateNetworkInterface',
              'ec2:CreateTags',
              'ec2:DeleteInstanceConnectEndpoint',
            ],
            Effect: 'Allow',
            Resource: '*',
          },
        ],
        Version: '2012-10-17',
      },
      Roles: [
        {
          Ref: Match.stringLikeRegexp('AWS.*ServiceRole.*'),
        },
      ],
    }));
  });

  it('Should have Custom', () => {
    template.hasResource('Custom::AWS', Match.objectEquals({
      Type: 'Custom::AWS',
      DeletionPolicy: 'Delete',
      UpdateReplacePolicy: 'Delete',
      DependsOn: [
        'EC2InstanceConnectCustomResourceCustomResourcePolicyF9AB3C7C',
      ],
      Properties: {
        Create: JSON.stringify({
          service: 'EC2',
          action: 'createInstanceConnectEndpoint',
          parameters: {
            SubnetId: 'example-subnet-id',
            DryRun: false,
            PreserveClientIp: true,
            SecurityGroupIds: [
              'example-security-group-id',
            ],
          },
          physicalResourceId: {
            responsePath: 'InstanceConnectEndpoint.InstanceConnectEndpointId',
          },
        }),
        Update: JSON.stringify({
          service: 'EC2',
          action: 'createInstanceConnectEndpoint',
          parameters: {
            SubnetId: 'example-subnet-id',
            DryRun: false,
            PreserveClientIp: true,
            SecurityGroupIds: [
              'example-security-group-id',
            ],
          },
          physicalResourceId: {
            responsePath: 'InstanceConnectEndpoint.InstanceConnectEndpointId',
          },
        }),
        Delete: JSON.stringify({
          service: 'EC2',
          action: 'deleteInstanceConnectEndpoint',
          parameters: {
            DryRun: false,
            InstanceConnectEndpointId: 'PHYSICAL:RESOURCEID:',
          },
        }),
        InstallLatestAwsSdk: true,
        ServiceToken: {
          'Fn::GetAtt': [
            'AWS679f53fac002430cb0da5b7982bd22872D164C4C',
            'Arn',
          ],
        },
      },
    }));
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});