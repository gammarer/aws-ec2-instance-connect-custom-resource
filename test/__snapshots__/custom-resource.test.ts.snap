// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomResource Testing Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "6bbce4c77566d179835e0a6de3293a27d9021f2f49d0055bb8e465eb6e1d89b1.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EC2InstanceConnectCustomResourceB58716F0": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "EC2InstanceConnectCustomResourceCustomResourcePolicyF9AB3C7C",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"EC2\\",\\"action\\":\\"createInstanceConnectEndpoint\\",\\"parameters\\":{\\"SubnetId\\":\\"example-subnet-id\\",\\"DryRun\\":false,\\"PreserveClientIp\\":true,\\"SecurityGroupIds\\":[\\"example-security-group-id\\"]},\\"physicalResourceId\\":{\\"responsePath\\":\\"InstanceConnectEndpoint.InstanceConnectEndpointId\\"}}",
        "Delete": "{\\"service\\":\\"EC2\\",\\"action\\":\\"deleteInstanceConnectEndpoint\\",\\"parameters\\":{\\"DryRun\\":false,\\"InstanceConnectEndpointId\\":\\"PHYSICAL:RESOURCEID:\\"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"EC2\\",\\"action\\":\\"createInstanceConnectEndpoint\\",\\"parameters\\":{\\"SubnetId\\":\\"example-subnet-id\\",\\"DryRun\\":false,\\"PreserveClientIp\\":true,\\"SecurityGroupIds\\":[\\"example-security-group-id\\"]},\\"physicalResourceId\\":{\\"responsePath\\":\\"InstanceConnectEndpoint.InstanceConnectEndpointId\\"}}",
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "EC2InstanceConnectCustomResourceCustomResourcePolicyF9AB3C7C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iam:CreateServiceLinkedRole",
                "ec2:CreateInstanceConnectEndpoint",
                "ec2:CreateNetworkInterface",
                "ec2:CreateTags",
                "ec2:DeleteInstanceConnectEndpoint",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EC2InstanceConnectCustomResourceCustomResourcePolicyF9AB3C7C",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
